[24.1] Tighten axt sniffer

martenson opened the pull request:
e.g. it would incorrectly sniff chain as axt since the comment line starts with "##matrix=axtChain"


## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

martenson committed:
tighten the axt format sniffer

e.g. it would incorrectly sniff chain as axt since the comment line starts with "##matrix=axtChain"

mvdbeek commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -1206,25 +1206,24 @@ def sniff_prefix(self, file_prefix: FilePrefix) -> bool:
+        headers = get_headers(file_prefix, None, count=4, comment_designator="#")
+        if not (
+            len(headers) == 4
+            and len(headers[0]) == 9
```
Could we instead do this ?
```suggestion
            and headers[0] == "matrix=axt"
```
that seems even more specific

martenson replied:
It would, but it is not a part of the standard as defined by https://genome.ucsc.edu/goldenPath/help/axt.html

Some tools add that comment line, others don't

mvdbeek replied:
Doesn't that mean you can't make that check at all, since now you're relying on this line having 9 characters, which feels fairly random ?

mvdbeek replied:
Ah, this is a non-comment line, nevermind.

bernt-matthias reviewed (commented):
Would it be an option to add a test case for the case that was a false positive before?

nsoranzo commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -1206,25 +1206,24 @@ def sniff_prefix(self, file_prefix: FilePrefix) -> bool:
+            and headers[0][2].isdigit()
+            and headers[0][3].isdigit()
+            and headers[0][5].isdigit()
+            and headers[0][6].isdigit()
```
`isdecimal()` seems to be better than `isdigit()` for all these.

martenson replied:
I agree, amended in [6ac51e5](https://github.com/galaxyproject/galaxy/pull/18204/commits/6ac51e5a9b5d8fb8265770349b181149f267bdc2), we might want to revisit the other sniffers doing this

nsoranzo commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -1206,25 +1206,24 @@ def sniff_prefix(self, file_prefix: FilePrefix) -> bool:
+        if not (
+            len(headers) == 4
+            and len(headers[0]) == 9
+            and headers[0][0].isdigit()
```
According to the standard:

```suggestion
            and headers[0][0] == "0"
```

nsoranzo commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -1206,25 +1206,24 @@ def sniff_prefix(self, file_prefix: FilePrefix) -> bool:
         """
-        headers = get_headers(file_prefix, None)
-        if len(headers) < 4:
+        headers = get_headers(file_prefix, None, count=4, comment_designator="#")
```
Can't you have an axt file of just 3 lines in theory?

martenson replied:
haha, I just pushed a fix for just that 🦅👀

martenson committed:
add unit test to sniffer, minor improvements

martenson committed:
tighten more!

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

martenson committed:
switch from isdigit to isdecimal

martenson committed:
switch to isdecimal

martenson committed:
refactor chain and ucsc.net sniffing

properly ignore comments
do not read the dataset unbounded, I think we have enough info from the headers and the first data line

martenson committed:
sorty sort
